zboxcore/sdk/writemarker_mutex.go

```diff
@@ -60,257 +52,215 @@ func CreateWriteMarkerMutex(client *client.Client, allocationObj *Allocation) (*
-	if m == nil {
-		return errors.Throw(constants.ErrInvalidParameter, "WriteMarkerMutex")
-	}
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-
-	if m.allocationObj == nil {
-		return errors.Throw(constants.ErrInvalidParameter, "allocationObj")
-	}
+func (wmMu *WriteMarkerMutex) Unlock(
+	ctx context.Context, mask zboxutil.Uint128,
+	blobbers []*blockchain.StorageNode,
+	timeOut time.Duration, connID string,
+) {
+	wg := &sync.WaitGroup{}
+	var pos uint64
+	for i := mask; !i.Equals64(0); i = i.And(zboxutil.NewUint128(1).Lsh(pos).Not()) {
+		pos = uint64(i.TrailingZeros())
 
-	T := len(m.allocationObj.Blobbers)
+		blobber := blobbers[pos]
 
-	if T == 0 {
-		return errors.Throw(constants.ErrInvalidParameter, "blobbers")
+		wg.Add(1)
+		go wmMu.UnlockBlobber(ctx, blobber, connID, timeOut, wg)
 	}
 
-	urls := make([]string, T)
-
-	builder := &strings.Builder{}
-	for i, b := range m.allocationObj.Blobbers {
-		builder.Reset()
-		builder.WriteString(strings.TrimRight(b.Baseurl, "/")) //nolint: errcheck
-		builder.WriteString(blobber.EndpointWriteMarkerLock)
-		builder.WriteString(m.allocationObj.Tx)
-		urls[i] = builder.String()
-	}
-
-	//protocol detail is on https://github.com/0chain/blobber/wiki/Features-Upload#upload
-	M := int(math.Ceil(float64(T) / float64(3) * float64(2))) //the minimum of M blobbers must accpet the marker
+	wg.Wait()
+}
 
-	//retry 3 times
-	for retry := 0; ; retry++ {
-		i := 1
-		n := 0
-		m.lockedBlobbers = nil
+// Change status code to 204
+func (wmMu *WriteMarkerMutex) UnlockBlobber(
+	ctx context.Context, b *blockchain.StorageNode,
+	connID string, timeOut time.Duration, wg *sync.WaitGroup,
+) {
 
-		body := url.Values{}
-		body.Set("connection_id", connectionID)
+	defer wg.Done()
 
-		now := time.Now()
-		body.Set("request_time", strconv.FormatInt(now.Unix(), 10))
+	var err error
 
-		form := body.Encode()
+	defer func() {
+		if err != nil {
+			logger.Logger.Error(err)
+		}
+	}()
 
-		for {
+	var req *http.Request
+	req, err = zboxutil.NewWriteMarkerUnLockRequest(
+		b.Baseurl, wmMu.allocationObj.Tx, connID, "")
+	if err != nil {
+		return
+	}
 
-			// M locks are acquired, it is safe to commit write
-			if n >= M {
-				return nil
-			}
+	var resp *http.Response
+	for retry := 0; retry < 3; retry++ {
+		if resp != nil {
+			resp.Body.Close()
+		}
 
-			// No more blobber, but n < M
-			if i > T && n < M {
-				//fails, release all locks
-				err := m.Unlock(ctx, connectionID)
-				if err != nil {
-					return err
-				}
-				break
-			}
+		reqCtx, ctxCncl := context.WithTimeout(ctx, timeOut)
+		resp, err = zboxutil.Client.Do(req.WithContext(reqCtx))
+		ctxCncl()
 
-			blobberIndex := i - 1
+		if err != nil {
+			return
+		}
+		if resp.StatusCode == http.StatusNoContent || resp.StatusCode == http.StatusOK {
+			logger.Logger.Info(b.Baseurl, connID, " unlocked")
+			return
+		}
 
-			result, err := m.lockOne(ctx, strings.NewReader(form), urls[blobberIndex])
+		if resp.StatusCode == http.StatusTooManyRequests {
```

Comment:
This retry code for StatusTooManyRequests is used in lot of places, it would be better to wrap the retry code into a common function so that could be reused.